text/0000-glimmer-components.md

```diff
@@ -0,0 +1,1022 @@
+
+### Properties
+
+Glimmer components have 3 properties: `args`, `isDestroying`, and `isDestroyed`.
+
+#### `args`
+
+As discussed in the motivation section, `args` is an object with the values of
+the named arguments passed to the component. This property will be updated
+whenever the arguments change, with the `didUpdateArgs` hook being called after
+any updates. It will be shallowly frozen in development mode to prevent users
+from setting values on it.
+
+#### `isDestroying`
+
+This property will be set to `true` when component teardown has been initiated,
+_before_ the component's `willDestroy` hook is run, along with any other
+components which are currently being torn down. This allows the entire component
+tree to be marked before user code is run. It can be used by users to
+conditionally prevent asynchronous code from running, and to check on the
+teardown state of the component in general.
+
+#### `isDestroyed`
+
+This property will be set after any `willDestroy` hooks have run, and the
+component has been fully torn down. It can be used by users to conditionally
+prevent asynchronous code from running, and to check on the teardown state of
+the component in general.
+
+### Lifecycle Hooks
+
+Classic components have 13 major lifecycle hooks that run during 3 major phases
+of the component lifecycle, with some hooks running during multiple phases:
+
+1. **Init and Initial Render:**
+   * `init`
+   * `willInsertElement`
+   * `didInsertElement`,
+   * `didReceiveAttrs`
+   * `willRender`
+   * `didRender`.
+
+2. **Rerenders and Updates:**
+   * `didReceiveAttrs`
+   * `didUpdateAttrs`
+   * `willUpdate`
+   * `didUpdate`
+   * `willRender`
+   * `didRender`
+
+3. **Destruction:**
+   * `willDestroyElement`
+   * `didDestroyElement`
+   * `destroy`
+   * `willDestroy`
+
+Many of these hooks have overlapping or redundant functionality, and it's fairly
+confusing when to use which and what the differences are. We can simplify this
+cycle in a number of ways:
+
+* Hooks that run during multiple phases such as `didRender` and
+  `didRecieveAttrs` can be convenient at times, but also add mental overhead and
+  redundancy. We can remove these in favor of clearly delineated hooks which
+  only run during _one_ phase.
+
+* "Bookend" methods (`did*` and `will*`) can be confusing, since they require
+  some specific knowledge of what the "bookended" functionality is. For
+  instance, users almost _always_ want to use `didInsertElement` and
+  `willDestroyElement`, but the existence of their opposite bookends can make
+  this confusing. Additionally, the fact that `didReceiveAttrs` and
+  `didUpdateAttrs` do _not_ have opposing bookends is inconsistent with this
+  pattern.
+
+* Hooks which are used to manipulate elements or the DOM in general can be
+  removed in favor of element modifiers, which are discussed in detail in the
+  next section.
+
+Based on these considerations, we can reduce these hooks to one per major phase
```

Comment:
Taking component instantiation out of the container may solve the problem, but the existence of `destroy()` on an object is an implicit protocol in a few places IIRC. We should make sure someone who has a `destroy()` action on a Glimmer component doesn't inadvertently get it called when the component is destroyed.

(You can close this comment, just an FYI with no impact on the document. 😄 )

Reply:
FYI - @pzuraq and I chatted through this quite a bit, destruction should _only_ be done by the component manager.

Below is a strawman that shows how we can ensure that the proper guarantees  are met without having to explicitly add a `destroy()` method:

```js
export default class Component {
  // other hooks and stuff here

  constructor(owner, args) {
    setOwner(this, owner);
    this.args = args;
    this.isDestroying = false;
    this.isDestroyed = false;
  }

  willDestroy() {}
}


export default class Manager {
  // ...snip...
  destroyComponent(bucket) {
    let { component } = bucket;
    let m = peekMeta(component);
   
    m.setSourceDestroying();
    component.isDestroying = true;

    schedule('actions', component, component.willDestroy);
    schedule('destroy', component, scheduledDestroy, component, m);
  }
}

function scheduledDestroy(component, m) {
  if (m.isSourceDestroyed()) {
    return;
  }
  m.destroy();
  m.setSourceDestroyed();
  component.isDestroyed = true;
}
```